### DISCUSSION TITLE
Nuxt 3 SSR Implementation

### DISCUSSION THREAD

**Discussion by Rigo-m** [algolia/instantsearch#5922] [Show and tell] [answered]
As stated in another discussion, here's my way of implementing instantsearch with Nuxt 3 and SSR
https://github.com/nuxt-modules/algolia/issues/187
For anyone needing it in the future.

As also stated in the linked issue, I don't think the implementation is optimal, I'd like to discuss on how to improve it using a more isomorphic approach

**Comment by Haroenv** [comment]
We're aware that the `this` is indeed harder to find in Vue 3 and we never found a good alternative for `script setup`. Do you have a solution in mind to retrieve the widgets on the server that works everywhere?

**Reply by Rigo-m** [reply]
I think we shouldn't tap into `this` at all.
Using composables and ssrContext to share state instead of mixins and dependency injection via provide/injection is the way to go.
Each widget should append its state to a shared composable (think of it as a `pinia` state or nuxt's `useState`). When each component gets transformed into a script setup one, and when each state gets saved into a composable unique state, you don't need `this` anymore, and you will just be able to have an isomorphic code where everything will work in ssr and client mode. State gets then serialized like it is now and hydrated if already present before mounting.
Ditching vue2 and options api can solidify and ease out the approach at the same time. Vue 2 will reach EOL in 2023, so I think it's ok to take it out of next versions of vue-instantsearch

**Reply by Haroenv** [reply]
In our usage we still 2/3 of our recent users using Vue 2, so deprecating it on our side likely isn't as straightforward. Thanks for your ideas though!

**Reply by Rigo-m** [reply]
Making a vue-instantsearch-composition package could be a long term solution? A-la react hooks.
Because right now, it really feels like nuxt 3 is a gimmick and not a real supported framework

**Reply by Rigo-m** [reply]
And then when vue 2 adoption falls to zero, vue-instantsearch could be deprecated in favour of composition support

**Reply by Haroenv** [reply]
We could consider this, but I'd like to note that in our stats, despite having long-time supported Vue, it's by far lower used than js or React. This causes it to usually slide lower in our priorities if that makes sense.

However if there's a way to simplify / improve what we have, we're obviously open to suggestions.

**Reply by Rigo-m** [reply]
If I produce a semi-advance starting point as a PR for the composition package, would you be willing to guide me through implementation when I'm stuck with some instnatsearch logic?

**Reply by Haroenv** [reply, marked as answer]
if you see an approach that is backwards-compatible (eg a feature flag without too much changes) and works with both v2 and v3, and are mainly stuck with tests etc., sure, I'm looking forward to see it!

**Reply by Youhan** [reply]
> ...it's by far lower used than js or React. This causes it to usually slide lower in our priorities if that makes sense.

Hi. Do we have any updates in 2025 stats? @Haroenv
Maybe "one" reason you don't see significant Vue3 / Nuxt3 / Nuxt4 users in the stats, is that there is no good support for Nuxt SSR?

**Reply by Rigo-m** [reply]
@Youhan in the mean time I have created a nuxt 3/4 ssr algolia instantsearch module. 
https://nuxt-swiftsearch.netlify.app/

**Reply by Youhan** [reply]
> @Youhan in the mean time I have created a nuxt 3/4 ssr algolia instantsearch module. 
> https://nuxt-swiftsearch.netlify.app/

Hi Matteo. Great job. I sent you an email earlier this week.
